piet-direct2d/src/dwrite.rs

```diff
@@ -133,42 +151,114 @@ impl DwriteFactory {
-    }
+impl FontCollection {
+    pub(crate) fn font_family(&self, name: &str) -> Option<FamilyName> {
+        let wname = name.to_wide_null();
+        let mut idx = u32::max_value();
+        let mut exists = 0_i32;
+
+        let family = unsafe {
+            let hr = self.0.FindFamilyName(wname.as_ptr(), &mut idx, &mut exists);
+            if SUCCEEDED(hr) && exists != 0 {
+                let mut family = null_mut();
+                let hr = self.0.GetFontFamily(idx, &mut family);
+                wrap(hr, family, FontFamily).ok()
+            } else {
+                eprintln!(
+                    "failed to find family name {}: err {} not_found: {}",
+                    name, hr, !exists
+                );
+                None
+            }
+        }?;
 
-    pub fn size(mut self, size: f32) -> TextFormatBuilder {
-        self.size = Some(size);
-        self
+        family.family_name().ok()
     }
+}
+
+impl FontFamily {
+    // this monster is taken right out of the docs :/
+    /// Returns the localized name of this family.
+    fn family_name(&self) -> Result<FamilyName, Error> {
+        unsafe {
+            let mut names = null_mut();
+            let hr = self.0.GetFamilyNames(&mut names);
+            if !SUCCEEDED(hr) {
+                return Err(hr.into());
+            }
+
+            let names: ComPtr<IDWriteLocalizedStrings> = ComPtr::from_raw(names);
+
+            let mut index = 0_u32;
+            let mut exists = 0_i32;
+            let mut locale_name = [0_u16; LOCALE_NAME_MAX_LENGTH];
+
+            let success =
+                GetUserDefaultLocaleName(locale_name.as_mut_ptr(), LOCALE_NAME_MAX_LENGTH as i32);
+            let mut hr = if SUCCEEDED(success) {
+                names.FindLocaleName(locale_name.as_ptr(), &mut index, &mut exists)
+            } else {
+                // we reuse the previous success; we want  to run the next block
+                // both if the previous failed, or if it just doesn't find anything
+                hr
+            };
+            if !SUCCEEDED(hr) || exists == 0 {
+                hr = names.FindLocaleName(DEFAULT_LOCALE.as_ptr(), &mut index, &mut exists);
+            }
+
+            if !SUCCEEDED(hr) {
+                return Err(hr.into());
+            }
+
+            // if locale doesn't exist, just choose the first
+            if exists == 0 {
+                index = 0;
+            }
+
+            let mut length = 0_u32;
+            let hr = names.GetStringLength(index, &mut length);
 
-    pub fn family(mut self, family: String) -> TextFormatBuilder {
-        self.family = Some(family);
-        self
+            if !SUCCEEDED(hr) {
+                return Err(hr.into());
+            }
+
+            let mut wide_name: Vec<u16> = Vec::with_capacity(length as usize + 1);
+            let hr = names.GetString(index, wide_name.as_mut_ptr(), length + 1);
+            if SUCCEEDED(hr) {
+                wide_name.set_len(length as usize + 1);
```

Comment:
I think maybe we want `set_len(length)` here so we don't include the trailing null.

Reply:
I think the null is part of the 'string', and should be included?

Reply:
On second thought, I think you're right. It's probably within the contract for the vec not to preserve the null.